components/tabs/src/storage.rs

```diff
@@ -268,44 +268,61 @@ impl TabsStorage {
         );
-        // Make a hash map of all urls per client_id that we potentially want to filter
-        let pending_closures = match pending_tabs_result {
-            Ok(pending_closures) => pending_closures.into_iter().fold(
-                HashMap::new(),
-                |mut acc: HashMap<String, Vec<String>>, (device_id, url)| {
-                    acc.entry(device_id).or_default().push(url);
-                    acc
-                },
-            ),
+        let pending_commands = match pending_commands_result {
+            Ok(pending_commands) => pending_commands,
             Err(e) => {
-                error_support::report_error!("tabs-read-remote", "Failed to read database: {}", e);
+                error_support::report_error!(
+                    "filter_pending_remote_commands",
+                    "Failed to read database: {}",
+                    e
+                );
                 return crts;
             }
         };
+
+        // Run over the result to build the commands in different structures:
+        // * A hash map of all urls which have been requested to be closed.
+        let mut pending_closures: HashMap<String, Vec<String>> = HashMap::new();
+        // * A simple vec of devices we want to "close all inactive" - building this in to a hashset seems a little over the top.
+        let mut pending_close_inactive = Vec::new();
+        for (device_id, command) in pending_commands {
+            match command {
+                RemoteCommand::CloseTab { url } => {
+                    pending_closures.entry(device_id).or_default().push(url)
+                }
+                RemoteCommand::CloseAllInactiveTabs => pending_close_inactive.push(device_id),
+            }
+        }
+
         // Check if any of the client records that were passed in have urls that the user closed
         // This means that they requested to close those tabs but those devices have not yet got
         // actually closed the tabs
         let filtered_crts: Vec<ClientRemoteTabs> = crts
             .into_iter()
             .map(|mut crt| {
+                let hide_inactive = pending_close_inactive.contains(&crt.client_id);
                 crt.remote_tabs.retain(|tab| {
-                    !pending_closures
-                        .get(&crt.client_id)
-                        // The top level in the url_history is the "active" tab, which we should use
-                        // TODO: probably not the best way to url check
-                        .map_or(false, |urls| urls.contains(&tab.url_history[0]))
+                    (!hide_inactive || !tab.inactive)
+                        && !pending_closures
+                            .get(&crt.client_id)
+                            // The top level in the url_history is the "active" tab, which we should use
+                            // TODO: probably not the best way to url check
+                            .map_or(false, |urls| urls.contains(&tab.url_history[0]))
```

Comment:
Woof, this entire expression took me a couple of tries to parse! IIUC, we're trying to exclude tabs that are either:

* Inactive, _only if_ there's a "close all inactive tabs" command pending for that device.
* Closed from this device, using the "close tabs" command.

Is that right?

In a follow-up—no need to do that here 😅—I wonder if we can do some refactoring, to make this a little easier to follow. Conceptually, it makes perfect sense, but I think the parallel data structures and conditionals obscure the flow a little.